Add HA Postgres features, with auto-failover/recovery

davidlohle opened the pull request:
This PR implements the functionality necessary for a single IP-based HA Postgres instance. It accomplishes this through HAProxy and a VRRP VIP and some health-check scripts. Essentially, it does as follows (excerpt from @jhunt)

```
On bootstrap, if there is no data directory, the postgres job will
revert to a normal, index-based setup.  The first node will assume
the role of the master, and the second will become a replica.

Once the data directory has been populated, future restarts of the
postgres job will attempt to contact the other node to see if it
is a master.  If the other node responds, and reports itself as a
master, the local node will attempt a `pg_basebackup` from the
master and assume the role of a replica.

If the other node doesn't respond, or reports itself as a replica,
the local node will keep trying, for up to
`postgres.replication.grace` seconds, at which point it will
assume the mantle of leadership and become the master node,
using its current data directory as the canonical truth.

Each node then starts up a `monitor` process; this process is
responsible for ultimately promoting a local replica to be a
master, in the event that the real master goes offline.  It works
like this:

  1. Busy-loop (via 1-second sleeps) until the local postgres
     instance is available on its configured port.  This prevents
     monitor from trying to restart the postgres while it is
     running a replica `pg_basebackup`.

  2. Busy-loop (again via 1-second sleeps) for as long as the
     local postgres is a master.

  3. Busy-loop (again via 1-second sleeps), checking the master
     status of the other postgres node, until it detects that
     either the master node has gone away (via a connection
     timeout), or the master node has somehow become a replica.

  4. Promote the local postgres node to a master.
```

This has been tested by hand in vSphere with CF's ccdb, uaa, and diegodb using postgres as it's storage medium. We noticed a `x < 5` second outage during master failover. Release notes are already written.

proplex committed:
Add some work

proplex committed:
Attempt at HA Postgres (failing miserably)

jhunt committed:
HAproxy approach that WORKS

jhunt committed:
README for HA

jhunt committed:
Default for connection timeout

jhunt committed:
Condense and simplify haproxy + keepalived into vip job

jhunt committed:
Default port for replication mode

jhunt committed:
Limit postgres instances to 2 when replicating

jhunt committed:
Upload blobs for my friends Dave and Patrick T. Jones

proplex committed:
Fix single-instance port being 6432

proplex committed:
Add readonly port forwarding through haproxy

proplex committed:
Update manifests and write release notes

jhunt reviewed (approved):
:+1:

daviddob commented on jobs/vip/templates/config/haproxy.conf:
```diff
@@ -0,0 +1,62 @@
+
+frontend postgres-read-replica
+mode tcp
+bind :<%= p('postgres.replication.readonly_port') || p('port') + 2000 %>
```
`postgres.replication.readonly_port` is intended to be optional, but if not specified, a deployment will fail here. It is also missing from the `vip` job spec file (tested with bosh-lite v264.7.0)

davidlohle replied:
Interesting, I never added it to the deployment `.yml` and it had worked/deployed fine. You're right about the specfile. Willfix.

daviddob replied:
Hmm, it refused to deploy till I added the following:

```
- name: vip
    properties:
      vip: ((vip_ip))
      postgres:
        replication:
          readonly_port: 7432
```

Corresponding output log:
```
Task 42 | 19:36:21 | Error: Unable to render instance groups for deployment. Errors are:
  - Unable to render jobs for instance group 'postgres'. Errors are:
    - Unable to render templates for job 'vip'. Errors are:
      - Error filling in template 'haproxy.conf' (line 53: Can't find property '["postgres.replication.readonly_port"]')
```

davidlohle replied:
If it fails, it should be pulling from the `|| p('port') + 2000`. I'll run some more tests and figure out why that's failing.

daviddob replied:
You may have to use an `if_p('postgres.replication.readonly_port')` or maybe use `p("postgres.replication.readonly_port", default_value)` instead of the `||` or define it in the spec file.

As per the docs for `p("some.property")` - If `some.property` is not supplied and does not have a default in the spec file, error will be raised to the user specifying that property value is missing

davidlohle replied:
Roger, thanks for pointing that out. I'll have all these tickets fixed soon.

jhunt replied:
Don't use `if_p`; the problem is that the spec doesn't have a definition for the property, which is a larger problem.  You can also set the default value in the second argument of `p()`, so you actually want this:

```
bind :<%= p('postgres.replication.readonly_port', p('port') + 2000) %>
```

(I think; try it and see what happens)

daviddob commented on jobs/postgres/templates/bin/monitor:
```diff
@@ -0,0 +1,85 @@
+  opts="${opts} -c hba_file=${JOB_DIR}/config/hba.conf"
+  opts="${opts} -c ident_file=${JOB_DIR}/config/ident.conf"
+
+  ln -sf ${JOB_DIR}/config/pg-master.conf \
```
Where does pg-master.conf come from? During testing when promote is called this leads to a broken link and postgres fails to start the next time.

davidlohle replied:
Old code, can be removed

daviddob commented on config/blobs.yml:
```diff
@@ -1,13 +1,28 @@
+  size: 601873
+  object_id: 09205505-e300-42b0-6513-24d4c4388960
+  sha: a69b2c40627a9bd69698a57e81a8c8c97826025d
+pgpool2/pgpool-II-3.7.4.tar.gz:
```
Is `pgpool2` a required package? I don't see any references to it anymore.

davidlohle replied:
Nope, can remove. Willfix.

daviddob reviewed (changes requested):
When testing what would occur if the nodes lost network connectivity to one another (but the instances themselves stayed running), it appears that in that case both nodes will become master and upon regaining connection, will continue to function independently.

Other than that, just a few other small bugs I encountered while testing a quick deployment.

jhunt commented:
Split-brain is almost impossible to prevent without a quorum solution, or some sort of binary star pattern.  We are mostly trying to protect against outages during a deployment.

proplex committed:
Adjust from feedback

In response to PR comments, I've made the following changes:

* Removed pgpool packages, as pgpool is no longer used in this release
* Removed references to `pg-master.conf`, code leftover from an earlier
attempt at HA-ifying this release
* Fixed `readonly_port` references, including moving it from
`postgres.replication.readonly_port` to `vip.readonly_port` (as this is
a haproxy related option, rather than a postgres option)
* Removed a extraneous postgres configuration logic that determined what
  interface to listen on.

davidlohle commented:
Feedback addressed!
